src/ray/core_worker/task_event_buffer.cc

```diff
@@ -0,0 +1,193 @@
+                                       report_interval_ms,
+                                       "CoreWorker.deadline_timer.flush_task_events");
+  return Status::OK();
+}
+
+void TaskEventBufferImpl::Stop() {
+  RAY_LOG(INFO) << "Shutting down TaskEventBuffer.";
+
+  // Shutting down the io service to exit the io_thread. This should prevent
+  // any other callbacks to be run on the io thread.
+  io_service_.stop();
+  if (io_thread_.joinable()) {
+    RAY_LOG(DEBUG) << "Joining io thread from TaskEventBuffer";
+    io_thread_.join();
+  }
+
+  {
+    absl::MutexLock lock(&mutex_);
+    // It's now safe to disconnect the GCS client since it will not be used by any
+    // callbacks.
+    if (gcs_client_) {
+      // Stop GCS client
+      gcs_client_->Disconnect();
+    }
+
+    gcs_client_.reset();
+  }
+}
+
+void TaskEventBufferImpl::AddTaskEvents(rpc::TaskEvents task_events) {
+  absl::MutexLock lock(&mutex_);
+
+  auto limit = RayConfig::instance().task_events_max_num_task_events_in_buffer();
+  if (limit > 0 && buffer_.size() >= static_cast<size_t>(limit)) {
+    // Too many task events, start overriding older ones.
+    buffer_[next_idx_to_overwrite_] = std::move(task_events);
+    next_idx_to_overwrite_ = (next_idx_to_overwrite_ + 1) % limit;
+    num_task_events_dropped_++;
+    return;
+  }
+  buffer_.push_back(std::move(task_events));
+}
+
+void TaskEventBufferImpl::FlushEvents(bool forced) {
+  std::vector<rpc::TaskEvents> task_events;
+  size_t num_task_events_dropped = 0;
+  {
+    absl::MutexLock lock(&mutex_);
+
+    RAY_LOG_EVERY_MS(INFO, 15000)
+        << "Pushed task state events to GCS. [total_bytes="
+        << (1.0 * total_events_bytes_) / 1024 / 1024
+        << "MiB][total_count=" << total_num_events_
+        << "][total_task_events_dropped=" << num_task_events_dropped_
+        << "][cur_buffer_size=" << buffer_.size() << "].";
+
+    // Skip if GCS hasn't finished processing the previous message.
+    if (grpc_in_progress_ && !forced) {
+      RAY_LOG_EVERY_N_OR_DEBUG(WARNING, 100)
+          << "GCS hasn't replied to the previous flush events call (likely overloaded). "
+             "Skipping reporting task state events and retry later."
+          << "[cur_buffer_size=" << buffer_.size() << "].";
+      return;
+    }
+
+    if (buffer_.size() == 0) {
+      return;
+    }
+
+    task_events.reserve(
+        RayConfig::instance().task_events_max_num_task_events_in_buffer());
+    buffer_.swap(task_events);
+    next_idx_to_overwrite_ = 0;
+    num_task_events_dropped = num_task_events_dropped_;
+    num_task_events_dropped_ = 0;
+  }
+  // mutex released. Below operations should be lock-free.
+
+  // Merge multiple events from a single task attempt run into one task event.
+  absl::flat_hash_map<std::pair<std::string, int>, rpc::TaskEvents> task_events_map;
+  for (auto events : task_events) {
+    auto &task_events_itr =
+        task_events_map[std::make_pair(events.task_id(), events.attempt_number())];
```

Comment:
can you add a test case for this? You can check the exported events have the expected keys?